epochStart/metachain/auctionListSelector.go

```diff
@@ -86,167 +176,252 @@ func (als *auctionListSelector) SelectNodesFromAuctionList(validatorsInfoMap sta
 		return nil
 	}
 
-	auctionListSize := uint32(len(auctionList))
-	log.Info("systemSCProcessor.SelectNodesFromAuctionList",
+	log.Info("auctionListSelector.SelectNodesFromAuctionList",
 		"max nodes", maxNumNodes,
 		"current number of validators", currNumOfValidators,
 		"num of nodes which will be shuffled out", numOfShuffledNodes,
 		"num of validators after shuffling", numOfValidatorsAfterShuffling,
 		"auction list size", auctionListSize,
-		fmt.Sprintf("available slots (%v -%v)", maxNumNodes, numOfValidatorsAfterShuffling), availableSlots,
+		fmt.Sprintf("available slots (%v - %v)", maxNumNodes, numOfValidatorsAfterShuffling), availableSlots,
 	)
 
-	err = als.sortAuctionList(auctionList, randomness)
-	if err != nil {
-		return err
-	}
-
+	als.displayOwnersData(ownersData)
 	numOfAvailableNodeSlots := core.MinUint32(auctionListSize, availableSlots)
-	als.displayAuctionList(auctionList, numOfAvailableNodeSlots)
-
-	for i := uint32(0); i < numOfAvailableNodeSlots; i++ {
-		newNode := auctionList[i]
-		newNode.SetList(string(common.SelectedFromAuctionList))
-		err = validatorsInfoMap.Replace(auctionList[i], newNode)
-		if err != nil {
-			return err
-		}
-	}
 
-	return nil
-}
+	sw := core.NewStopWatch()
+	sw.Start("auctionListSelector.sortAuctionList")
+	defer func() {
+		sw.Stop("auctionListSelector.sortAuctionList")
+		log.Debug("time measurements", sw.GetMeasurements()...)
+	}()
 
-// TODO: Move this in elrond-go-core
-func safeSub(a, b uint32) (uint32, error) {
-	if a < b {
-		return 0, epochStart.ErrUint32SubtractionOverflow
-	}
-	return a - b, nil
+	return als.sortAuctionList(ownersData, numOfAvailableNodeSlots, validatorsInfoMap, randomness)
 }
 
-func getAuctionListAndNumOfValidators(validatorsInfoMap state.ShardValidatorsInfoMapHandler) ([]state.ValidatorInfoHandler, uint32) {
-	auctionList := make([]state.ValidatorInfoHandler, 0)
+func (als *auctionListSelector) getAuctionDataAndNumOfValidators(
+	validatorsInfoMap state.ShardValidatorsInfoMapHandler,
+	unqualifiedOwners map[string]struct{},
+) (map[string]*ownerData, uint32, uint32, error) {
+	ownersData := make(map[string]*ownerData)
 	numOfValidators := uint32(0)
+	numOfNodesInAuction := uint32(0)
 
 	for _, validator := range validatorsInfoMap.GetAllValidatorsInfo() {
-		if validator.GetList() == string(common.AuctionList) {
-			auctionList = append(auctionList, validator)
+		blsKey := validator.GetPublicKey()
+		owner, err := als.stakingDataProvider.GetBlsKeyOwner(blsKey)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+
+		if isInAuction(validator) {
+			_, isUnqualified := unqualifiedOwners[owner]
+			if isUnqualified {
+				log.Debug("auctionListSelector: found node in auction with unqualified owner, do not add it to selection",
+					"owner", hex.EncodeToString([]byte(owner)),
+					"bls key", hex.EncodeToString(blsKey),
+				)
+				continue
+			}
+
+			err = als.addOwnerData(owner, validator, ownersData)
```

Comment:
I think you should start from data from the stakingProvider component and add the necessary information. You need to know the keys which are in the auction list for this owner and the total topup and other global info about the owner.

Reply:
Yes, refactored all code in a separate PR: https://github.com/ElrondNetwork/elrond-go/pull/4155/files